### DISCUSSION TITLE
Gamepad/keyboard navigation support

### DISCUSSION THREAD

**Discussion by MAGIKru** [organicmaps/organicmaps#1690] [Ideas]
Android devices are able to work with connected via OBD/Bluetooth gamepads and keyboards.
After some investigation I didn't find any navigation app that has fully support of attached controllers.

My idea is very simple. Attach some keys events (user should be able to remap default keys) to screen gestures like these:
Minimum - most important:
1. Arrows should move the map up/down/left/right
2. A/B buttons should zoom in/out
3. X button should center map to the current location
Additional:
4. Switch/toggle layers
5. Rotate and reset compas

Why this feature is important? 
Many of us use map applications as navigation while driving. In these case very often the phone is mounted in most visible place, but it may be not easy to access with the fingers. In case when user drive motorcycle - it also is not safe to drive using only one hand, while with the second one you move the map and try to understand where to drive next. Especially when user has gloves and touch screen doesn't see touches. As a solution - users can buy small and cheap gamepads from aliexpress, that could be mounted very close to the the handlebar grips, and navigation will become more safety and comfortable.

The second scenario - is to use android device as screen mirroring on the TV and have possibility to control the map using wireless keyboard.

**Comment by biodranik** [comment]
Interesting idea, thanks ) However, there are some questions:
1. Why driver needs to drag/zoom the map while driving? It is not safe! A properly implemented navigation should solve this. If it doesn't, let's improve the existing implementation.
2. Not everything can be mapped. Search, for example. How to solve that?
3. For home TV, why do you need the map on it, with limited control? For screen mirroring, for example, you may control the map from the device directly.

**Comment by MAGIKru** [comment]
1.1 In my case - I often use map without navigation started wile drive motorcycle on offroad trails. I this case I use just map mode to see where are the rivers with bridges and the possible trails around me, after that I visually analyse difficulty of these trails and decide where to go next. Sometimes they are dead-end path. Using map moving I can prevent unneeded movements where I will have to turn back. Permanent usage of small zoom is also not comfortable, sometimes trails are close to each other and I need increase the zoom level. And the main problem as I told - are thick gloves. So, navigation mode improvement is not a solution while offroad throw the forest or fields. 
1.2 Another case - while drive on the public roads and navigation - sometimes is useful to check and analyse another possible roads to drive when you see that there is a traffic jam or an accident or repairs ahead. Or to check route/turn ahead with other friends, that are using another navigation apps/devices (different devices have different notification about next turn, some of them ignore lite turns, but some notify, and while driving in group we have to verify if everybody have the same route long ahead). 
2.1 Using gamepad controller - search is not important imho. As option - to have additional button hotkey to start navigation to the central point (cursor) on the map. So, user will move the map to the needed place, then press the special button and start navigation to that point. But I think this solution will be much difficult to implement than simple attaching key events to movement and zoom.
2.2 Using keyboard - I see possible solution is to set hotkey to open search box, then user type query as simple text (other action events are paused, until user finish search using Enter or Esc key).
3. As example: I have connected TV/Projector using cable connection (Type C adapted). On the map I have several favourites points. I want to have a briefing to present my route suggestion. Control the map using device directly is possible, but not always comfortable due to cable length limitation and TV/Projector placement. Will be much comfortable to have wireless gamepad/keyboard controller and sit in the desk.

As I said - no one popular map apps that I have tested don't have such universal controllers support. Usually they use default system switch focus to the interface buttons elements like TAB button, but it is not useful at all.

**Reply by biodranik** [reply]
I don't think that it's hard to add controller support. I assume it should not break any existing controls. So any contribution is welcome!
